core/server/common/src/main/java/alluxio/master/journal/JournalTool.java

```diff
@@ -239,4 +214,469 @@ private static void usage() {
+      serializer.resolve(new StorageSerialization());
+
+      SingleThreadContext context = new SingleThreadContext("readJournal", serializer);
+
+      try {
+        // Read through the whole journal content, starting from snapshot.
+        context.execute(this::readCopycatSnapshotFromDir).get();
+        context.execute(this::readCopycatLogFromDir).get();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw e;
+      } catch (ExecutionException e) {
+        throw e.getCause();
+      }
+    }
+
+    private void readCopycatLogFromDir() {
+      try (PrintStream out =
+          new PrintStream(new BufferedOutputStream(new FileOutputStream(mJournalEntryFile)))) {
+        Log log = Storage.builder().withDirectory(mInputDir).build().openLog("copycat");
+        for (long i = log.firstIndex(); i < log.lastIndex(); i++) {
+          io.atomix.copycat.server.storage.entry.Entry entry = log.get(i);
+          if (entry instanceof CommandEntry) {
+            Command command = ((CommandEntry) entry).getCommand();
+            if (command instanceof JournalEntryCommand) {
+              byte[] entryBytes = ((JournalEntryCommand) command).getSerializedJournalEntry();
+              try {
+                writeSelected(out, JournalEntry.parseFrom(entryBytes));
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            }
+          }
+        }
+      } catch (Exception e) {
+        LOG.error("Failed to read from journal.", e);
+      }
+    }
+
+    private void readCopycatSnapshotFromDir() {
+      Storage journalStorage = Storage.builder().withDirectory(mInputDir).build();
+      if (journalStorage.openSnapshotStore("copycat").snapshots().isEmpty()) {
+        LOG.debug("No snapshot found.");
+        return;
+      }
+      Snapshot currentSnapshot = journalStorage.openSnapshotStore("copycat").currentSnapshot();
+
+      String snapshotDumpFile =
+          String.format("copycat-%s-%s.log", currentSnapshot.index(), currentSnapshot.timestamp());
+
+      try (CheckpointInputStream checkpointStream =
+          new CheckpointInputStream(new SnapshotReaderStream(currentSnapshot.reader()))) {
+        readCheckpoint(checkpointStream, Paths.get(mCheckpointsDir, snapshotDumpFile));
+      } catch (Exception e) {
+        LOG.error("Failed to read from journal.", e);
+      }
+    }
+
+    private NetworkAddressUtils.ServiceType getRaftServiceType() {
+      if (mMaster.contains("job")) {
+        return NetworkAddressUtils.ServiceType.JOB_MASTER_RAFT;
+      } else {
+        return NetworkAddressUtils.ServiceType.MASTER_RAFT;
+      }
+    }
+
+    /**
+     * Writes given entry after going through range and validity checks.
+     *
+     * @param out out stream to write the entry to
+     * @param entry the entry to write to
+     */
+    private void writeSelected(PrintStream out, JournalEntry entry) {
```

Comment:
Probably out of scope of this PR - we should consider consolidate methods writing journal entries for both raft journal and ufs journal.

Reply:
For UFS it's pretty straightforward because the journal could be selected for a master.
For RAFT, we need to do filtering, because we have the whole journal system coming in the same stream.